### PR TITLE
feat: support DTS autoExtension

### PR DESCRIPTION
**Pull request by Timeless0911**
## Summary

Support DTS autoExtension
- `.d.ts` with esm format and `.d.cts` with cjs format when `type: module`
- `.d.ts` with cjs format and `.d.mts` with esm format when `type: commonjs`

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/config.ts [hunk 1]
```diff
@@ -405,7 +410,7 @@ async function composeLibRsbuildConfig(
     config.bundle,
     dirname(configPath),
   );
-  const dtsConfig = await composeDtsConfig(config);
+  const dtsConfig = await composeDtsConfig(config, dtsExtension);
```
Comment (fi3ework) [inline packages/core/src/config.ts:413]: Could we move the dts logic in `composeAutoExtensionConfig` into `composeDtsConfig`, then DTS config will not couple with JS anymore.

File: packages/plugin-dts/src/apiExtractor.ts [hunk 2]
```diff
@@ -5,23 +5,34 @@ import {
...
     entry = 'index.d.ts',
     tsconfigPath = 'tsconfig.json',
   } = options;
   try {
-    const untrimmedFilePath = join(cwd, relative(cwd, outDir), 'index.d.ts');
+    const start = Date.now();
+    const untrimmedFilePath = join(
+      cwd,
+      relative(cwd, outDir),
+      `index${dtsExtension}`,
```
Comment (fi3ework) [inline packages/plugin-dts/src/apiExtractor.ts:34]: The `index` is fixed for now?

Comment (Timeless0911) [reply packages/plugin-dts/src/apiExtractor.ts:34]: What should our bundled DTS file named

File: packages/core/src/config.ts (hunk 1 above)
Comment (Timeless0911) [reply packages/core/src/config.ts:413]: I aim to reduce multiple time call for `getDefaultExtension`
image
 Or is this acceptable?

File: packages/plugin-dts/src/apiExtractor.ts (hunk 2 above)
Comment (fi3ework) [reply packages/plugin-dts/src/apiExtractor.ts:34]: Provide naming config to user, we could implement this in the future.

File: packages/core/src/config.ts (hunk 1 above)
Comment (fi3ework) [reply packages/core/src/config.ts:413]: So be it like this, we could do some refactor in the future.

**Review by fi3ework — APPROVED** [review]

**Pull request merged**